### ISSUE TITLE
[inductor] performance regression after change to Grid2D

### ISSUE THREAD

**Issue by jianyizh** [pytorch/pytorch#177117]
I see models in torchbench like basic_gnn_gin has performance regression compare to 2.10 on 4080 super （0.37ms -> 0.59ms). 
`python benchmarks/dynamo/torchbench.py --inference --float32 -d xpu -n10 --performance --disable-cudagraphs --only basic_gnn_gin --cold-start-latency --backend=inductor`

triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4 increase from 0.077ms to 0.15ms. The difference is it changes from Grid1D to 2D

current triton kernel

```python
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

from torch._dynamo.testing import rand_strided
from torch._C import _cuda_getCurrentRawStream as get_raw_stream
import torch

@triton_heuristics.pointwise(
    size_hints={'y': 262144, 'x': 64}, tile_hint=TileHint.DEFAULT,
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*i64', 'in_ptr1': '*fp32', 'in_ptr2': '*fp32', 'out_ptr0': '*fp32', 'ynumel': 'i32', 'xnumel': 'i32', 'YBLOCK': 'constexpr', 'XBLOCK': 'constexpr'}, 'device': DeviceProperties(type='cuda', index=0, multi_processor_count=80, cc=89, major=8, regs_per_multiprocessor=65536, max_threads_per_multi_processor=1536, max_threads_per_block=1024, warp_size=32), 'constants': {}, 'native_matmul': False, 'enable_fp_fusion': True, 'launch_pdl': False, 'disable_ftz': False, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]], (3,): [['tt.divisibility', 16]], (4,): [['tt.divisibility', 16]], (5,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid2DWithYZOverflow', 'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4', 'mutated_arg_names': ['out_ptr0'], 'optimize_mem': True, 'no_x_dim': False, 'atomic_add_found': True, 'num_load': 3, 'num_store': 1, 'num_reduction': 0, 'backend_hash': 'B48C87BAA4B2EAD66C986D3E7384F474EF1F365401B51326FDA68405CF23A049', 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'deterministic': False, 'force_filter_reduction_configs': False, 'mix_order_reduction_allow_multi_stages': False, 'are_deterministic_algorithms_enabled': False, 'tiling_scores': {'y': 3200000, 'x': 256}, 'kernel_num_gb': 0.008320256, 'kernel_flop': 0},
    min_elem_per_thread=0
)
@triton.jit
def triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4(in_ptr0, in_ptr1, in_ptr2, out_ptr0, ynumel, xnumel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr):
    ynumel = 200000
    xnumel = 64
    yoffset = (tl.program_id(1) + tl.program_id(2) * tl.num_programs(1)) * YBLOCK
    yindex = yoffset + tl.arange(0, YBLOCK)[:, None]
    ymask = yindex < ynumel
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[None, :]
    xmask = xindex < xnumel
    y0 = yindex
    x1 = xindex
    tmp0 = tl.load(in_ptr0 + (200000 + y0), ymask, eviction_policy='evict_last')
    tmp2 = tl.load(in_ptr0 + (y0), ymask, eviction_policy='evict_last')
    tmp8 = tl.load(in_ptr1 + (x1), xmask, eviction_policy='evict_last')
    tl.device_assert(((0 <= tmp0) & (tmp0 < 10000)) | ~(ymask), "index out of bounds: 0 <= tmp0 < 10000")
    tmp3 = tl.full([1, 1], 10000, tl.int32)
    tmp4 = tmp2 + tmp3
    tmp5 = tmp2 < 0
    tmp6 = tl.where(tmp5, tmp4, tmp2)
    tl.device_assert(((0 <= tmp6) & (tmp6 < 10000)) | ~(ymask), "index out of bounds: 0 <= tmp6 < 10000")
    tmp9 = tl.load(in_ptr2 + (x1 + 64*tmp6), xmask & ymask)
    tmp10 = tmp8 + tmp9
    tmp11 = tl.full([1, 1], 0, tl.int32)
    tmp12 = triton_helpers.maximum(tmp11, tmp10)
    tl.atomic_add(out_ptr0 + (tl.broadcast_to(x1 + 64*tmp0, [YBLOCK, XBLOCK])), tmp12, xmask & ymask, sem='relaxed')


def get_args():
    arg_0 = rand_strided((2, 200000), (200000, 1), device='cuda:0', dtype=torch.int64)
    arg_1 = rand_strided((64,), (1,), device='cuda:0', dtype=torch.float32)
    arg_2 = rand_strided((10000, 64), (64, 1), device='cuda:0', dtype=torch.float32)
    arg_3 = rand_strided((10000, 64), (64, 1), device='cuda:0', dtype=torch.float32)
    return arg_0, arg_1, arg_2, arg_3, 200000, 64,


def call(args):
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        stream0 = get_raw_stream(0)
        triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4.run(*args, stream=stream0)

def benchmark_all_configs(args):
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        return triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4.benchmark_all_configs(*args)


if __name__ == '__main__':
    from torch._inductor.runtime.benchmarking import benchmarker

    args = get_args()
    ms = benchmarker.benchmark(lambda: call(args), device='cuda', rep=40)
    num_gb = 0.008320256
    gb_per_s = num_gb / (ms / 1e3)
    print(f"{ms:.3f}ms    {num_gb:.3f}GB    {gb_per_s:.2f}GB/s")
```

old triton kernel
```python
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

from torch._dynamo.testing import rand_strided
from torch._C import _cuda_getCurrentRawStream as get_raw_stream
import torch

@triton_heuristics.pointwise(
    size_hints={'x': 16777216},
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*i64', 'in_ptr1': '*fp32', 'in_ptr2': '*fp32', 'out_ptr0': '*fp32', 'xnumel': 'i32', 'XBLOCK': 'constexpr'}, 'device': DeviceProperties(type='cuda', index=0, multi_processor_count=80, cc=89, major=8, regs_per_multiprocessor=65536, max_threads_per_multi_processor=1536, max_threads_per_block=1024, warp_size=32), 'constants': {}, 'native_matmul': False, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]], (3,): [['tt.divisibility', 16]], (4,): [['tt.divisibility', 16]]}], 'enable_fp_fusion': True},
    inductor_meta={'grid_type': 'Grid1D', 'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4', 'mutated_arg_names': ['out_ptr0'], 'optimize_mem': True, 'no_x_dim': False, 'atomic_add_found': True, 'num_load': 3, 'num_store': 1, 'num_reduction': 0, 'backend_hash': 'B48C87BAA4B2EAD66C986D3E7384F474EF1F365401B51326FDA68405CF23A049', 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'deterministic': False, 'force_filter_reduction_configs': False, 'are_deterministic_algorithms_enabled': False, 'kernel_num_gb': 0.008320256, 'kernel_flop': 0},
    min_elem_per_thread=0
)
@triton.jit
def triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4(in_ptr0, in_ptr1, in_ptr2, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 12800000
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)[:]
    x1 = xindex // 64
    x0 = (xindex % 64)
    tmp0 = tl.load(in_ptr0 + (200000 + x1), None, eviction_policy='evict_last')
    tmp2 = tl.load(in_ptr0 + (x1), None, eviction_policy='evict_last')
    tmp8 = tl.load(in_ptr1 + (x0), None, eviction_policy='evict_last')
    tl.device_assert((0 <= tmp0) & (tmp0 < 10000), "index out of bounds: 0 <= tmp0 < 10000")
    tmp3 = tl.full([XBLOCK], 10000, tl.int32)
    tmp4 = tmp2 + tmp3
    tmp5 = tmp2 < 0
    tmp6 = tl.where(tmp5, tmp4, tmp2)
    tl.device_assert((0 <= tmp6) & (tmp6 < 10000), "index out of bounds: 0 <= tmp6 < 10000")
    tmp9 = tl.load(in_ptr2 + (x0 + 64*tmp6), None)
    tmp10 = tmp8 + tmp9
    tmp11 = tl.full([1], 0, tl.int32)
    tmp12 = triton_helpers.maximum(tmp11, tmp10)
    tl.atomic_add(out_ptr0 + (x0 + 64*tmp0), tmp12, None, sem='relaxed')


def get_args():
    arg_0 = rand_strided((2, 200000), (200000, 1), device='cuda:0', dtype=torch.int64)
    arg_1 = rand_strided((64,), (1,), device='cuda:0', dtype=torch.float32)
    arg_2 = rand_strided((10000, 64), (64, 1), device='cuda:0', dtype=torch.float32)
    arg_3 = rand_strided((10000, 64), (64, 1), device='cuda:0', dtype=torch.float32)
    return arg_0, arg_1, arg_2, arg_3, 12800000,


def call(args):
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        stream0 = get_raw_stream(0)
        triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4.run(*args, stream=stream0)


def benchmark_all_configs(args):
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        return triton_poi_fused_addmm_expand_index_select_new_zeros_relu_scatter_add_select_view_4.benchmark_all_configs(*args)


if __name__ == '__main__':
    from torch._inductor.runtime.benchmarking import benchmarker

    args = get_args()
    ms = benchmarker.benchmark(lambda: call(args), device=cuda, rep=40)
    num_gb = 0.008320256
    gb_per_s = num_gb / (ms / 1e3)
    print(f"{ms:.3f}ms    {num_gb:.3f}GB    {gb_per_s:.2f}GB/s")

```

cc @jerryzh168 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by jianyizh** [comment]
Hi @eellison. I have updated the kernel in description. Sorry for pasting wrong kernel. Since this kernel has atomic add, the performance is also depend on the value of arg_0, so I run the whole model on 4080 super and check the kernel performance by Nsight. Looks like grid2d sectors/Req is lower and it uses shared memory. However, we are expecting better coalesce for 2d.
grid1d:

<img width="655" height="155" alt="Image" src="https://github.com/user-attachments/assets/6d6db53f-2550-43ce-b5bc-4102964ed84e" />

<img width="680" height="145" alt="Image" src="https://github.com/user-attachments/assets/9baad26c-b0a7-48c8-b6dd-32e1c6a648ec" />

 grid2d:

<img width="650" height="152" alt="Image" src="https://github.com/user-attachments/assets/367d48bf-8ccc-4342-8919-db397b0162ed" />

<img width="662" height="145" alt="Image" src="https://github.com/user-attachments/assets/3e8d842b-4629-421b-b63d-870b5399aca6" />

**Comment by eellison** [comment]
@jianyizh are you interested in submitting fix ?

**Comment by jianyizh** [comment]
> [@jianyizh](https://github.com/jianyizh) are you interested in submitting fix ?

@eellison Sorry, I don't know how to fix it yet.

**Comment by razaaliraza** [comment]
The root cause is in `analyze_memory_coalescing` in `tiling_utils.py`: `coalesced_by_var` records which variable coalesces each memory expression, but not whether that coalescing depends on the tiling choice. Broadcast loads (stride 0 wrt the innermost pointwise variable) and already-coalesced loads (stride 1) are already efficient in 1D, but their scores still inflate 2D candidate totals.

The fix filters these scores before they enter `coalesced_by_var`. For each expression attributed to a non-innermost variable, a two-point stride check determines if the access is already coalesced in 1D. If so, the score is not added. Conservative default: when the stride is non-constant or undeterminable, the score is kept (identical to current behavior).

Benchmarked broadcast outer product and scatter kernels on H200, which has enough scheduling throughput to partially mask the grid overhead. The tiling choice change (2D to 1D) is consistently reproduced, but the performance difference is smaller than the 2x reported on RTX 4080 Super. The 1-5% improvement on H200 is a lower bound; the regression is more pronounced on consumer hardware with fewer SMs.

```
                  name  before (2D)  after (1D)  improvement
--------------------------------------------------------------
         bcast 1M x 16     22.09 us    21.07 us        4.6%
         bcast 4M x 16     72.23 us    71.42 us        1.1%
        bcast 10M x 16    182.63 us   178.54 us        2.2%
        bcast 20M x 16    366.96 us   352.70 us        3.9%
        bcast 50M x 16    914.30 us   877.93 us        4.0%
       bcast 100M x 16   1827.89 us  1749.15 us        4.3%
         bcast 10M x 8     96.34 us    93.26 us        3.2%
         bcast 20M x 8    191.41 us   182.83 us        4.5%
         bcast 50M x 8    474.21 us   451.51 us        4.8%
        bcast 100M x 8    948.49 us   899.15 us        5.2%
      scatter 200K f64     25.68 us    25.10 us        2.3%
         scatter 8M f8    145.80 us   145.87 us        0.0%
```

PR incoming. Let me know what you think. Thanks!
